src/Data/Discrimination/Internal.hs

```diff
@@ -55,3 +65,28 @@ spanEither k xs0 = go [] xs0 where
 fromRight :: Either a b -> b
 fromRight (Right y) = y
 fromRight _ = error "unstable discriminator"
+
+-------------------------------------------------------------------------------
+-- * Integer and Natural
+-------------------------------------------------------------------------------
+
+integerCases :: Integer -> Either (GmpSize,[GmpLimb]) (Either Int (GmpSize,[GmpLimb]))
+integerCases (Jn# b) = Left          $ decomposeBigNat b
+integerCases (S#  i) = Right . Left  $ I# i
+integerCases (Jp# b) = Right . Right $ decomposeBigNat b
+{-# INLINE integerCases #-}
+
+naturalCases :: Natural -> Either GmpLimb (GmpSize,[GmpLimb])
+naturalCases (NatS# w) = Left $ W# w
+naturalCases (NatJ# b) = Right $ decomposeBigNat b
+{-# INLINE naturalCases #-}
+
+-- We need to reverse the limb array. Its stored least-significant word first
+-- but for comparasion to work right we need most-significant words first.
+decomposeBigNat :: BigNat -> (GmpSize, [GmpLimb])
+decomposeBigNat (BN# ba) = let pa = PrimArray ba :: PrimArray Word in (sizeofPrimArray pa, primArrayToReverseList pa)
+{-# INLINE decomposeBigNat #-}
+
+primArrayToReverseList :: Prim a => PrimArray a -> [a]
+primArrayToReverseList xs = build (\c n -> foldlPrimArray (flip c) n xs)
```

Comment:
Does this have the same effect as the following? It looks like it still needs to iterate over all elements of the list before producing the first element 
```haskell
foldrPrimArrayReverse :: forall a b. Prim a => (a -> b -> b) -> b -> PrimArray a -> b
foldrPrimArrayReverse f z arr = go (sz-1)
  where
    !sz = sizeofPrimArray arr
    go !i
      | i <= 0 = f (indexPrimArray arr i) (go (i-1))
      | otherwise = z
```

(or even simply
```haskell
primArrayToReverseList : Prim a => PrimArray a -> [a]
primArrayToReverse xs = go (sizeofPrimArray arr - 1)
  where
    go i | i < 0 = []
         | otherwise = indexPrimArray xs i : go (i-1)
```

Reply:
I took `primArrayToList` and swapped `foldr` to `foldl`. Won't do any more "optimizations" before you show they are worth the extra code complexity (benchmarks!)

Reply:
Oh I missed the foldl! Ignore me